### PR TITLE
Add __module__ to torch.dtype.__dict__

### LINKED ISSUE #65077: Pickling `torch.dtype` objects is unnecessarily slow
## 🐛 Bug

Pickling `torch.dtype` objects is unnecessarily slow, due to forcing `pickle` to search the entire Python module namespace to find the right module.

## To Reproduce

A simple benchmark:

```
❯ python -m timeit -s 'import torch, pickle' 'pickle.dumps(torch.float32)'
1000 loops, best of 5: 212 usec per loop
```

Compare to pickling a built-in type, or the class `torch.dtype`:
```
❯ python -m timeit -s 'import torch, pickle' 'pickle.dumps(float)'
500000 loops, best of 5: 687 nsec per loop

❯ python -m timeit -s 'import torch, pickle' 'pickle.dumps(torch.dtype)'
500000 loops, best of 5: 626 nsec per loop
```

Notice the `usec` vs `nsec` -- `torch.float32` is around _3000_ times slower.

I believe the problem is that `torch.dtype` instances:
1. define a `__reduce__` which returns a string, and
2. lack a `__module__` property to tell Pickle the name of the module to which they belong.

In this case, Python reacts by (approximately) **searching every module in `sys.modules`** to find a module s.t. `getattr(mod, obj.__reduce__()) is obj`.

We can reproduce this fallback behavior with a simple script:

```
❯ cat bench.py
import timeit
import pickle
from dataclasses import dataclass

@dataclass
class FakeDtype:
  name: str

  def __reduce__(self):
    return self.name

int = FakeDtype('int')

if __name__ == '__main__':
  N = 10000
  t1 = timeit.timeit(lambda: pickle.dumps(int), number=N)
  print(f"Pickling our fake dtype: {1000000 * t1 / N:0.3f}us")
  int.__module__ = None
  t2 = timeit.timeit(lambda: pickle.dumps(int), number=N)
  print(f"Pickling our fake dtype (no __module__): {1000000 * t2 / N:0.3f}us")

❯ python -m bench
Pickling our fake dtype: 1.156us
Pickling our fake dtype (no __module__): 29.290us
```

We see a similar slowdown when we remove the `__module__` field.

## Expected behavior

I believe the correct fix here is to add a `__module__` attribute to the `torch.dtype` class, which contains the value `torch`. This will be inherited via its instances and avoid the `sys.modules` walk.

## Environment

```
PyTorch version: 1.9.0
Is debug build: False
CUDA used to build PyTorch: None
ROCM used to build PyTorch: N/A

OS: macOS 11.5.2 (arm64)
GCC version: Could not collect
Clang version: 12.0.5 (clang-1205.0.22.9)
CMake version: Could not collect
Libc version: N/A

Python version: 3.9.2 | packaged by conda-forge | (default, Feb 21 2021, 05:00:30)  [Clang 11.0.1 ] (64-bit runtime)
Python platform: macOS-11.5.2-arm64-arm-64bit
Is CUDA available: False
CUDA runtime version: No CUDA
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA
HIP runtime version: N/A
MIOpen runtime version: N/A

Versions of relevant libraries:
[pip3] mypy-extensions==0.4.3
[pip3] numpy==1.20.2
[pip3] pytorch-memlab==0.2.3
[pip3] torch==1.9.0
[pip3] torchtyping==0.1.1
[conda] Could not collect
```

## Additional context

cc @VitalyFedyunin @ngimel @heitorschueroff @mruberry

### PR DESCRIPTION
**Pull request by nelhage**
torch.dtype.__reduce__ returns a string, which causes Pickle to look
up the object by module and name. In order to find the right module,
Pickle looks for __module__ on the object; if it doesn't find that, it
falls back to searching sys.modules.

Previously, torch.dtype instances did not have a `__module__`
attribute, so pickling dtypes would fall back to a search of
sys.module.

Instances of normal Python objects have a `__module__` attribute
because normal Python classes have a `__module__` key in their
`__dict__`. Imitate that by populating one in `torch.dtype`.

We set the field in `tp_dict` before calling `PyType_Ready` (instead
of afterwards) because of the doc warning against mutating a type's
dictionary once initialized:
https://docs.python.org/3/c-api/typeobj.html#c.PyTypeObject.tp_dict

fixes #65077

---

I didn't add any tests because I didn't see any obvious places with similar tests for pickling or dtype objects. Let me know if I missed the right place, or should start one.

### CODE REVIEW

File: torch/csrc/Dtype.cpp [hunk 1]
```diff
@@ -126,6 +126,22 @@ PyTypeObject THPDtypeType = {
 
 void THPDtype_init(PyObject *module)
 {
+  // Set a __dict__ with `__module__` = `torch`. This means
+  // `__module__` value will be inherited by instances, which will
+  // prevent Pickle from having to search all of sys.modules in order
+  // to find the module when pickling a dtype.
+  //
+  // See https://github.com/pytorch/pytorch/issues/65077
+  AT_ASSERT(THPDtypeType.tp_dict == nullptr);
```
Comment (mruberry) [inline torch/csrc/Dtype.cpp:135]: AT_ASSERT is deprecated, use TORCH_CHECK instead

File: torch/csrc/Dtype.cpp [hunk 2]
```diff
@@ -126,6 +126,22 @@ PyTypeObject THPDtypeType = {
 
 void THPDtype_init(PyObject *module)
 {
+  // Set a __dict__ with `__module__` = `torch`. This means
```
Comment (mruberry) [inline torch/csrc/Dtype.cpp:129]: What does this look like in Python?

Like how does a user perceive this change on torch.dtype and torch.float32, for example?

File: torch/csrc/Dtype.cpp [hunk 3]
```diff
@@ -126,6 +126,22 @@ PyTypeObject THPDtypeType = {
...
+  // `__module__` value will be inherited by instances, which will
```
Comment (mruberry) [inline torch/csrc/Dtype.cpp:130]: This comment may need to be clarified. Like if we consider a Python program

```
class foo(object):

  def __init__(self):
    # self.__dict__ = {"__module__": "silly"}
    self.__module__ = "silly"

class bar(foo):

  def __init__(self):
    super().__init__()
```

then

```
b = bar()
b.__module__
: silly
```

so why not just set the dunder module attribute directly?

Comment (nelhage) [reply torch/csrc/Dtype.cpp:130]: For a C extension type, we can't set it directly in Python:
```
>>> torch.dtype.__module__ = 'torch'
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: can't set attributes of built-in/extension type 'torch.dtype'
>>> torch.dtype.__dict__['__module__'] = 'torch'
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: 'mappingproxy' object does not support item assignment
```

and doing a `PyObject_SetAttrString` I believe will fail for similar reasons.

File: torch/csrc/Dtype.cpp (hunk 2 above)
Comment (nelhage) [reply torch/csrc/Dtype.cpp:129]: The linked https://github.com/pytorch/pytorch/issues/65077 explains the ultimate motivation, but in short, before this. change, `torch.float32.__module__` does not exist, and after this change, it does. This effects the efficiency of pickling dtypes.

File: torch/csrc/Dtype.cpp (hunk 3 above)
Comment (mruberry) [reply torch/csrc/Dtype.cpp]: I see; so this is the recommended way of setting these attributes on C extension types?

Sorry I didn't respond earlier, I'm on vacation at the moment but am now checking notifications.

**Comment by mruberry** [comment]
This seems reasonable to me but being on PTO @ezyang would you shepherd this PR through? I think you're more familiar with these extension types, too.

**Review by ezyang — APPROVED** [review]
Man, CPython does not make life easy

File: torch/csrc/Dtype.cpp (hunk 3 above)
Comment (nelhage) [reply torch/csrc/Dtype.cpp]: I couldn't find a standard idiom here. Populating `tp_dict` is explicitly supported in the docs, and user-defined types also have a `__module__` in `__dict__` so I feel comfortable that all the components of this solution are standard in isolation, but I can't rule out that there could be another approach here.

**Pull request closed without merging**